### PR TITLE
add explicit-createref-type

### PR DESCRIPTION
**Pull request by meishuu**
In light of discordapp/discord#11250 I figured it'd be prudent to enforce this at linter level to ensure we handle it moving forward.

<sub>Also, I have no idea what I'm doing with writing custom ESLint rules.</sub>

(Could use assistance figuring out how to test this on our live codebase before merge.)

### CODE REVIEW

**Comment by joeybaker** [comment]
OMG. You read my mind!!!!

Also: I'm 80% done actually adding the types for `createRef` calls :)

File: lib/rules/explicit-createref-type.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+const explicitCreateRefTypeMessage = `React.createRef() class properties should be explicitly typed.`;
```
Comment (joeybaker) [inline lib/rules/explicit-createref-type.js:3]: What do you think of specifying that the type should be: `{current: null | React.ElementRef<typeof Component>}`?

File: yarn.lock [hunk 2]
```diff
@@ -2,6 +2,84 @@
 # yarn lockfile v1
 
 
+"@babel/code-frame@^7.0.0":
```
Comment (joeybaker) [inline yarn.lock:5]: Why did all this change?

**Review by joeybaker — COMMENTED** [review]
Code looks good, I'm concerned about the yarn.lock changes though.

**Comment by meishuu** [comment]
@joeybaker ESLint normally has no idea how to parse our class property syntax, so we need to pull in `babel-eslint` as the custom parser:

https://github.com/discordapp/eslint-plugin-discord/blob/1ec72f2119474e3c6f723997293c4e19e2d3b775/tests/lib/rules/explicit-createref-type.js#L4

This is only used for testing this new rule, so it should be pretty safe. The `ClassProperty` node type comes from `babel-eslint` so the rule no-ops if included in a project that doesn't support them (and ESLint will bail if using an unsupported parser on code that uses this syntax).

**Comment by joeybaker** [comment]
Ahh! I totally missed that in the `package.json`. Sorry!

**Review by joeybaker — APPROVED** [review]
Seems good. I think you can test this with `yarn link`?

File: lib/rules/explicit-createref-type.js (hunk 1 above)
Comment (meishuu) [reply lib/rules/explicit-createref-type.js:3]: Hm; not sure it's worth the extra code complexity in the rule right now. I can also envision a future where we might be shorthanding it à la `React$CreateRef<typeof Component>` or something.

I think we can land this for now and evaluate again after some time; if the rule isn't enough of a hint to developers, we can be more strict about the check.

**Comment by TheSisb** [comment]
To test on the main app:
- Open the `package.json` in the main app's root dir
- Find the `eslint-plugin-discord` module 
- Change the hash to `d855db6b0477bd42b30522301809dde4aee54194` (your latest commit)
- Run `yarn`
- (Optional) open the node_modules/eslint-plugin-discord folder to verify it got the files
- run our eslint command in the app

**Review by TheSisb — APPROVED** [review]
lgtm

**Review by joeybaker — APPROVED** [review]

**Pull request merged**